engine/tests/phpunit/Elgg/Notifications/NotificationsServiceTestCase.php

```diff
@@ -0,0 +1,929 @@
+		
+		return [
+			$object,
+			$group,
+			$user,
+			$metadata,
+			$annotation,
+			$relationship,
+		];
+	}
+
+	public function testRegisterEvent() {
+		$this->setupServices();
+
+		$object = $this->getTestObject();
+
+		$this->notifications->registerEvent($object->getType(), $object->getSubtype());
+		$events = array(
+			$object->getType() => array(
+				$object->getSubtype() => array('create')
+			)
+		);
+		$this->assertEquals($events, $this->notifications->getEvents());
+
+		$this->notifications->registerEvent($object->getType(), $object->getSubtype(), array('test_event'));
+		$events[$object->getType()][$object->getSubtype()] = array('create', 'test_event');
+		$this->assertEquals($events, $this->notifications->getEvents());
+
+		$this->notifications->registerEvent($object->getType(), $object->getSubtype());
+		$this->assertEquals($events, $this->notifications->getEvents());
+	}
+
+	public function testUnregisterEvent() {
+		$this->setupServices();
+
+		$object = $this->getTestObject();
+
+		$this->notifications->registerEvent($object->getType(), $object->getSubtype());
+		$this->assertTrue($this->notifications->unregisterEvent($object->getType(), $object->getSubtype()));
+
+		$events = array(
+			$object->getType() => array()
+		);
+		$this->assertEquals($events, $this->notifications->getEvents());
+		$this->assertFalse($this->notifications->unregisterEvent($object->getType(), $object->getSubtype()));
+	}
+
+	public function testRegisterMethod() {
+		$this->setupServices();
+
+		$this->notifications->registerMethod('test_method');
+		$methods = array('test_method' => 'test_method');
+		$this->assertEquals($methods, $this->notifications->getMethods());
+	}
+
+	public function testUnregisterMethod() {
+		$this->setupServices();
+
+		$this->notifications->registerMethod('test_method');
+		$this->assertTrue($this->notifications->unregisterMethod('test_method'));
+		$this->assertEquals(array(), $this->notifications->getMethods());
+		$this->assertFalse($this->notifications->unregisterMethod('test_method'));
+	}
+
+	public function testEnqueueEvent() {
+		$this->setupServices();
+
+		$object = $this->getTestObject();
+
+		$this->session->setLoggedInUser($this->actor);
+
+		$this->notifications->registerEvent($object->getType(), $object->getSubtype());
+
+		$this->notifications->enqueueEvent('create', $object->getType(), $object);
+
+		$event = new SubscriptionNotificationEvent($object, 'create');
+		$this->assertEquals($event, $this->queue->dequeue());
+		$this->assertNull($this->queue->dequeue());
+
+		// unregistered action type
+		$this->notifications->enqueueEvent('null', $object->getType(), $object);
+		$this->assertNull($this->queue->dequeue());
+
+		// unregistered object type
+		$this->notifications->enqueueEvent('create', $object->getType(), new ElggObject());
+		$this->assertNull($this->queue->dequeue());
+
+		$this->session->removeLoggedInUser();
+	}
+
+	public function testStoppingEnqueueEvent() {
```

Comment:
The more descriptive names, the better. This could be e.g. `testPluginHookHandlerCanStopEnqueueEvent()`.